### ISSUE TITLE
gumboc.py, soup_adapter, and python 3

### ISSUE THREAD

**Issue by kevinhendricks** [google/gumbo-parser#345]
Hi,
I have need to use gumboc.py with python3 along with soup_adapter.py.   I have noticed that the current gumboc.py and gumboc_test.py have mismatch issues with __str__ being called on bytes objects (which is forbidden in python3). 

I would like to create a version of gumboc.py and gumboc_test.py that works well on both python and python3.  Is there any interest in patches to make this happen?  If so, would you like the gumboc.py StringPieces tag names, original_tag_names, etc  to be returned as full unicode or as bytes objects.   Since gumbo only works with utf-8 encoded files, converting between them is trivial but I thought we would want both python 2 and python 3 to return full unicode for use in python instead of str on python 2 and bytes on python 3 to be consistent.

In other words ... this would involve add the following to each python script:

from __future__ import unicode_literals, division, absolute_import, print_function

import sys
PY3 = sys.version_info[0] == 3
if PY3:
    binary_type = bytes
    text_type = str
else:
    binary_type = str
    text_type = unicode

And then making conversions based on the if (type(text) == binary_type) so that the it works on both PY3 and PY2

Also, I have hacked up a version of soup_adapter to work with BeautifulSoup4 (bs4) on python3 and the older BeautifulSoup3 on python2.  Is there any interest in pull requests or patches for these changes?

Thanks,

Kevin

**Comment by nostrademons** [comment]
Yes, I'm interested in any patches related to Python3 compliance.  It had worked for a bit - I used Gumbo for a Python3 project a year or so back - but it's possible it regressed with some of the next/prev changes leading up to 0.10, or that there were other Python3 incompatibilities that weren't exposed by my particular use-case.  Tests and updates to travis.yml to make sure future regressions don't happen are also welcome.

On implementation - I agree that unicode is probably more appropriate than bytes.  c_char_p in ctypes gives you a bytes, unfortunately, and I think there's no way to get data out of the C struct other than c_char_p (if I'm wrong about this, feel free to use whatever mechanism will convert directly from utf-8 char* to unicode).  How about providing an @property accessor for these fields that calls .decode('utf-8') on the property?  I'd prefer that decoding happen at system boundaries, i.e. as soon as data leaves the C struct and enters Python, it gives a unicode object rather than have to test the type in multiple places.  That means you shouldn't have to do many (if any) type tests - str.decode('utf-8') gives you a unicode object in Python2, and bytes.decode('utf-8') gives you a str object in Python3, so just inserting the decode call should work properly in both versions.

On versioning - changing the return value of these fields is a backwards-incompatible change.  Couple options I'd be happy with:

1. Leave the existing field names and add @property accessors prefixed with 'unicode_' - so things like Node.unicode_tag, Attribute.unicode_value, etc.  This is backwards-compatible and can be merged into v0.10.2.
2. Change the existing field names to return unicode strings.  Since we need to have the ctypes fields anyway, the ctypes c_char_p fields would be named something like "bytes_name" or even "_name", and then there'd be an @property accessor to provide name as "self._name.decode('utf-8')".  I like this better from an API perspective, but it'd have to merge into 1.0.0 to preserve backwards compatibility.  soup_adapter.py and html5lib_adapter.py would also have to change, as they assume they're getting back bytes from gumboc.

I'll also take any patches to add a BS4 adapter; the only reason I never did one was lack of time.

**Comment by kevinhendricks** [comment]
@nostrademons

The more I look at this, the more I think the gumboc.py ctypes interface needs to only return byte strings at all times.  The main reason is  that all of the text/source positions and text lengths become confusing if mixing utf-8 encoded strings and unicode encoded strings with and offsets into the document text buffer, versus column information.

If we supply only a pure byte string based interface, at least people can follow along with the basic types and use only byte offsets and not have to worry.  

For example see the StringPiece class

    class StringPiece(ctypes.Structure):
      _fields_ = [
          ('data', _Ptr(ctypes.c_char)),
          ('length', ctypes.c_size_t),
          ]

      def __len__(self):
        return self.length

      def __str__(self):                                                            
        return ctypes.string_at(self.data, self.length)


The str operator overload to define the str for StringPiece actually fails on python 3 because it is an runtime error to ever have the str operator return a bytes object.  This is where the current gumboc_test.py fails on python 3 as it constantly uses str() on specific StringPiece objects.

Now if we fix this by conditioning on PY3, we end up with the str operator returning a byte string object on python 2.7 and a unicode (normal str) on python 3.x.

It gets even more confusing if we look at the len() operator for that same class.  Would we need to return the length in unicode chars for python 3.x and the number of bytes for python 2.X?

This quickly gets worse when you think about column values vs offsets.

I think the best solution is to consistently (across python versions) return bytes strings for all string values.  One way to do that for the StringPiece class is as follows:

    class StringPiece(ctypes.Structure):
      _fields_ = [
          ('data', _Ptr(ctypes.c_char)),
          ('length', ctypes.c_size_t),
          ]

      def __len__(self):
        return self.length

      def __bytes__(self):                                                            
        return ctypes.string_at(self.data, self.length) 

The nice thing is that "bytes" is an "alias" for str in python 2.X (after 2.5 or so I think) so that invoking str(StringPiece) on python 2.X is not broken and bytes() on the StringPiece class object would always be easily understood as utf-8 encoded bytestring data by all with a consistent len result.

The final user of the gumbc.py would see all string information as  bytes and if they really wanted a unicode  representation they could easily use a snippet of code to convert any object to its unicode string representation that would work on things like StringPiece object

    def unistr(obj):
        return bytes(obj).decode('utf-8')

So I think what already exists in gumboc.py, and gumbo_test.py just needs a few small tweaks to actually work on both python 2 and 3 with a single code-base.  The soup_adapter.py needs much more work in order to use bs4 on python 3.x but I have something that is working now  preliminarily.

As much as I would like pure unicode strings to be used throughout, I think it is just not easily doable while keeping such a thin ctypes wrapper around C structures in gumbo.  

 What do you think?  

Thanks,

Kevin

**Comment by kevinhendricks** [comment]
@nostrademons
Please see the following patch against master:

https://gist.github.com/kevinhendricks/2025df8d7161168d7ecc

With this patch in place on a clean build, I can do the following using python 2.7.X and python 3.4.X

KevinsiMac:gumbo kbhend$ python ./gumboc_test.py
......
----------------------------------------------------------------------
Ran 6 tests in 0.001s

OK

KevinsiMac:gumbo kbhend$ python3 ./gumboc_test.py
......
----------------------------------------------------------------------
Ran 6 tests in 0.001s

OK


KevinsiMac:gumbo kbhend$ python ./bs4_adapter.py

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
<html lang="en-US" xml:lang="en" xmlns="http://www.w3.org/1999/xhtml/"><head><title>testing &amp; entities</title></head>
<body>
  <p class="first second">this is the <i><b>copyright</b></i> symbol "©"</p>
  <p class="second" xlink:href="http://www.ggogle.com" xmlns:xlink="http://www.w3.org/xlink">this used to test atribute namespaces</p>


</body></html>
<head><title>testing &amp; entities</title></head>
<p class="first second">this is the <i><b>copyright</b></i> symbol "©"</p>
<p class="second" xlink:href="http://www.ggogle.com" xmlns:xlink="http://www.w3.org/xlink">this used to test atribute namespaces</p>


KevinsiMac:gumbo kbhend$ python3 ./bs4_adapter.py

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
<html lang="en-US" xml:lang="en" xmlns="http://www.w3.org/1999/xhtml/"><head><title>testing &amp; entities</title></head>
<body>
  <p class="first second">this is the <i><b>copyright</b></i> symbol "©"</p>
  <p class="second" xlink:href="http://www.ggogle.com" xmlns:xlink="http://www.w3.org/xlink">this used to test atribute namespaces</p>


</body></html>
<head><title>testing &amp; entities</title></head>
<p class="first second">this is the <i><b>copyright</b></i> symbol "©"</p>
<p class="second" xlink:href="http://www.ggogle.com" xmlns:xlink="http://www.w3.org/xlink">this used to test atribute namespaces</p>


KevinsiMac:gumbo kbhend$ python ./bs4_adapter_test.py
.
----------------------------------------------------------------------
Ran 1 test in 0.002s

OK


KevinsiMac:gumbo kbhend$ python3 ./bs4_adapter_test.py
.
----------------------------------------------------------------------
Ran 1 test in 0.001s

OK
